### PR TITLE
NODE-1525: refactor auth process to use new public API

### PR DESCRIPTION
**Pull request by mbroadst**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,9 +1,15 @@
 'use strict';
 
+const kerberos = require('./lib/kerberos');
+
 // Get the Kerberos library
-module.exports = require('./lib/kerberos');
+module.exports = kerberos;
+
+// Support legacy versions of the mongodb driver which expect this export
+module.exports.Kerberos = kerberos;
```
Comment (daprahamian) [inline index.js:9]: Optional: This circular reference will probably be bad for future module imports.Can we either:

A) deprecate this and come up with a plan for removing it?
B) make a ticket for doing `A`?

File: lib/auth_processes/mongodb.js [hunk 2]
```diff
@@ -1,383 +1,142 @@
 'use strict';
```
Comment (daprahamian) [inline lib/auth_processes/mongodb.js:1]: Lets make a ticket to move this into `mongodb-core`

File: lib/auth_processes/mongodb.js [hunk 3]
```diff
@@ -1,383 +1,142 @@
...
+  transition(payload, callback) {
+    if (this._transition == null) {
+      return callback(new Error('Transition finished'));
+    }
 
-UnixMongoProcessor.prototype.transition = function(payload, callback) {
-  if (this._transition == null) return callback(new Error('Transition finished'));
-  this._transition(payload, callback);
-};
+    this._transition(payload, callback);
```
Comment (daprahamian) [inline lib/auth_processes/mongodb.js:70]: This whole transition mechanic looks like another impl of a state machine. Can we add a todo/note to revisit this when we eventually come around to trying to make a single state machine impl?

**Review by daprahamian — APPROVED** [review]
LGTM. My comments are all either optional or TODOs for later.

**Review by kvwalker — APPROVED** [review]
looks good!

File: index.js (hunk 1 above)
Comment (mbroadst) [reply index.js:9]: I thought about this a bit last night. Unfortunately, this is the only way to do this without requiring users to update their driver. At the same time it's frustrating to make this change in a major release, as we will have to live with it. I created [NODE-1600](https://jira.mongodb.org/browse/NODE-1600) to deal with this.

File: lib/auth_processes/mongodb.js (hunk 2 above)
Comment (mbroadst) [reply lib/auth_processes/mongodb.js:1]: [NODE-1601](https://jira.mongodb.org/browse/NODE-1601)

File: lib/auth_processes/mongodb.js (hunk 3 above)
Comment (mbroadst) [reply lib/auth_processes/mongodb.js:70]: added to a comment in [NODE-1517](https://jira.mongodb.org/browse/NODE-1517)

**Pull request merged**